### DISCUSSION TITLE
GraphQL UI does not show exceptions that are thrown inside subscription

### DISCUSSION THREAD

**Discussion by Bazalii** [quarkusio/quarkus#30580] [Q&A] [answered]
After successful connection to server with subscription endpoint graphql UI in dev mode doesn't show exceptions that are thrown by subscription. Is it an expected behaviour? 

For example, I have created a subscription inside which an exception is thrown when given token is not `"123"` 

```
@Subscription
fun subscribeToUsersWithFilter(context: Context, token: String, userName: String): Multi<User> {
    if (token != "123"){
        throw CustomException("Invalid token!")
    }

    return _broadcastProcessor.filter { user -> user.name == userName }
}
```

But dev UI shows only that connection was successful and nothing else.
![image](https://user-images.githubusercontent.com/58232044/214341458-e9e7024d-0d90-4811-b3dc-79c2fe3d2b4a.png)

Furthermore, I don't see any error messages in my message history of websocket connection in browser
![image](https://user-images.githubusercontent.com/58232044/214366954-464226e2-c1c0-4ca8-9f62-cb507b918a52.png)

[a comment by an automated account is not shown]

**Comment by phillip-kruger** [comment, marked as answer]
Are there any errors in the browser's console  ? I don't think (even though the ui say it did) a connection has been established. The Error happens before that. Can you share a reproducer ?

**Reply by phillip-kruger** [reply]
We can probably move this discussion to SmallRye.

**Reply by Bazalii** [reply]
This is the [reproducer](https://github.com/Bazalii/Quarkus-GraphQL-Subscriptions).

First, send request for subscription with invalid "token":

```
subscription subscribeToUsersWithFilter($token: String!, $userName: String!) {
  subscribeToUsersWithFilter(token: $token, userName: $userName) {
    id
    name
    roles
  }
}
```
variables:
```
{
  "userName": "Test",
  "token": "12"
}
```

After that send mutation:
```
mutation addUser($user: UserInput!) {
    addUser(user: $user) {
    id
    name
    roles
  }
```

variables:
```
{
  "user": {
    "id": "00000000-0000-0100-0000-000000000000",
    "name": "Test",
    "roles": [
      "Worker",
      "User"
    ]
  }
}
```

Exception is thrown inside the application but nothing is send to the client.

Another thing here is If you comment line 37 in UserController, you will see that exception is thrown when the first User is added using mutation and it is sent after connection and processed correctly.

**Reply by jmartisk** [reply]
It seems we're not properly handling the situation that the subscription method itself throws an exception, instead of returning a failed `Multi`. Generally, if a method returns a `Multi`, it should always return it, and preferably an exception should be wrapped in it, instead of being thrown from the method itself (so, you return `Multi.createFrom().failure(exception)`. In that case the server sends a proper error message and ends the operation.

 But yes, the library should expect that this can happen. I'll look into fixing this.

**Reply by jmartisk** [reply]
I've submitted a PR that should help with this case: https://github.com/smallrye/smallrye-graphql/pull/1705
In any case I still recommend returning a failed Multi instead of throwing an exception directly.

**Reply by Bazalii** [reply]
Okay, thank you!
